climada/hazard/trop_cyclone.py

```diff
@@ -405,7 +405,19 @@ def _apply_criterion(self, criterion, scale):
                 change = chg['change'] * scale
             if select.any():
                 new_val = getattr(haz_cc, chg['variable'])
-                new_val[select] *= change
+                #For sparse matrices *= is ineficient (slow, uses large memory)
+                #Instead use multiply and chunking
+                if isinstance(new_val, sparse.csr_matrix):
+                    chunk_size = 10
+                    n_chunks = int(len(select) / chunk_size)
+                    chunks = np.arange(0, n_chunks, chunk_size)
+                    chunks = np.append(chunks, n_chunks)
+                    for start, stop in zip(chunks[:-1], chunks[1:]):
+                        new_val[select[start:stop]] = \
+                            new_val[select[start:stop]].multiply(change)
```

Comment:
First of all: Didn't you attend my talk about avoiding for-loops? Okay, just joking. :D

This code doesn't do what you think it does. I pushed new test cases to your branch that demonstrate this.

Let me quickly explain, what's going wrong: The matrix `select[start:stop]` is a binary matrix of size 10 (or smaller). With a dense NumPy array, this would raise an IndexError:
```
>>> import numpy as np
>>> import scipy.sparse as sparse
>>> select_start_stop = [True, False, True]
>>> new_val = np.arange(24).reshape(6,4)
>>> new_val
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])
>>> new_val[select_start_stop]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: boolean index did not match indexed array along dimension 0; dimension is 6 but corresponding boolean dimension is 3
```
However, with SciPy's sparse matrices, there is no error. The mask is assumed to be for the beginning of the axis and `False` is assumed as a value for the missing entries:
```
>>> sparse.csr_matrix(new_val)[select_start_stop].toarrray()
array([[ 0,  1,  2,  3],
       [ 8,  9, 10, 11]])
```
So, you are only ever indexing (and modifying) the first 10 rows of the sparse matrix which is definitely not what you intend, right?

Here is my proposed change:
```suggestion
                    new_val = new_val.multiply((change * select  + ~select)[:, None]).tocsr()
```

Reply:
I have another proposal that uses matrix multiplication which is even faster:
```suggestion
                    new_val = sparse.diags(np.where(select, change, 1)).dot(new_val)
```

Reply:
Aaaa yeah of course. Doing stuff late when tired is not a good idea :D. Somehow, although I immediately thought of the diagonal matrix multiplication, in the night I was convinced that the multiplication with a diagonal matrix would not give the right result. Thanks!